### ISSUE TITLE
Asynchronous i/o not working well

### ISSUE THREAD

**Issue by wstomv** [gibiansky/IHaskell#932]
Using the IPython kernel, executing the following code cell
``` python
print("Hello, what is your name: ", end="")
name = input()
print(name)
print("And what is your age: ", end="")
age = input()
print(age)
```
will result in an interactive dialog, where prompts, inputs, and responses are nicely interleaved.

However, trying the same thing with the IHaskell kernel fails:
``` haskell
dialog :: IO()
dialog = do
  putStr "Give me a string: "
  line <- getLine
  putStrLn line
  putStr "Give me another one: "
  line2 <- getLine
  putStrLn line2

dialog
```

With my locally installed IHaskell kernel (under MacOS; IHaskell 0.9.0.3 GHC 8.2.2), the input is read, _without_ any output being shown, and once cell execution is finished _all_ the output is shown together. I.e., i/o is not asynchronous.

With the IHaskell kernel on https://mybinder.org/v2/gh/gibiansky/IHaskell/master (IHaskell 0.9.1.0 GHC 8.2.2), you get this exception:

    <stdin>: hGetLine: end of file

Currently, this is a showstopper for our educational use of IHaskell.

**Comment by erikd** [comment]
This issue is actually totally unrelated to IHaskell, but actually due to the default buffering of `stdout` in Haskell (its probably line buffered). When you call `putStr` the runtime will wait for you to print a newline.

The solution is to add a `hFlush stdout` after each `putStr`, or maybe even define a function:
```
putStrFlush :: String -> IO ()
putStrFlush str = do
   putSrr str
   hFlush stdout
```
and use that instead.

**Comment by wstomv** [comment]
@erikd Indeed, it still does not work satisfactorily with `hFlush`.  But it is a step forward.
In fact, using `putStrLn` instead of `putStr` already helps in simple programs.

When there is more/slow output, the input box is briefly flashed on the screen (actually, you see this box before the preceding output was printed) and it then disappears. After this, the kernel hangs, and needs restarting.

**Comment by vaibhavsagar** [comment]
We [reimplement our IO handling](https://github.com/gibiansky/IHaskell/blob/83a823d5f98b1821ec158c527754631d10e1d40e/src/IHaskell/IPython/Stdin.hs), so maybe the problem is there?
